### PR TITLE
feat: add Crimson Nylium and Warped Nylium blocks

### PR DESCRIPTION
**Pull request by AzaleeX**
## Introduction
`Crimson Nylium` and `Warped Nylium` blocks added
image

## Tests

I tested this PR by doing the following (tick all that apply):
- [x] Writing PHPUnit tests (commit these in the `tests/phpunit` folder)
- [x] Playtesting using a Minecraft client (provide screenshots or a video)
- [x] Writing a test plugin (provide the code and sample output)
- [x] Other (provide details)

### CODE REVIEW

File: src/block/GrassNylium.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	public function ticksRandomly() : bool{
+		return true;
+	}
+
+	public function onRandomTick() : void{
+		$world = $this->position->getWorld();
+		$lightAbove = $world->getFullLightAt((int) $this->position->x, (int) $this->position->y + 1, (int) $this->position->z);
+		if($lightAbove < 4 && $world->getBlockAt((int) $this->position->x, (int) $this->position->y + 1, (int) $this->position->z)->getLightFilter() >= 2){
```
Comment (ShockedPlot7560) [inline src/block/GrassNylium.php]: Better to use `getSide(Facing::UP)`

File: src/block/NetherRoots.php [hunk 2]
```diff
@@ -30,11 +30,13 @@ final class NetherRoots extends Flowable{
...
 
 	private function canBeSupportedAt(Block $block) : bool{
-		//TODO: nylium, moss
+		//TODO: moss
 		$supportBlock = $block->getSide(Facing::DOWN);
 		return
 			$supportBlock->hasTypeTag(BlockTypeTags::DIRT) ||
 			$supportBlock->hasTypeTag(BlockTypeTags::MUD) ||
+			$supportBlock->getTypeId() === BlockTypeIds::CRIMSON_NYLIUM ||
+			$supportBlock->getTypeId() === BlockTypeIds::WARPED_NYLIUM ||
```
Comment (ShockedPlot7560) [inline src/block/NetherRoots.php]: Better to create a new Nylium tag for that is think

File: src/block/RedMushroom.php [hunk 3]
```diff
@@ -23,13 +23,11 @@
...
 
-use pocketmine\item\Item;
+use pocketmine\block\utils\StaticSupportTrait;
 use pocketmine\math\Facing;
-use pocketmine\math\Vector3;
-use pocketmine\player\Player;
-use pocketmine\world\BlockTransaction;
 
 class RedMushroom extends Flowable{
+	use StaticSupportTrait;
```
Comment (ShockedPlot7560) [inline src/block/RedMushroom.php:30]: Apply only minimal change, this is unrelated to the PR "add Crimson Nylium".

Open an other PR if you want to change that.

File: src/block/VanillaBlocks.php [hunk 4]
```diff
@@ -932,6 +934,9 @@ public function getBreakTime(Item $item) : float{
 		$grassBreakInfo = BreakInfo::shovel(0.6);
 		self::register("grass", new Grass(new BID(Ids::GRASS), "Grass", new Info($grassBreakInfo, [Tags::DIRT])));
 		self::register("grass_path", new GrassPath(new BID(Ids::GRASS_PATH), "Grass Path", new Info($grassBreakInfo)));
+		self::register("crimson_nylium", new GrassNylium(new BID(Ids::CRIMSON_NYLIUM), "Crimson Nylium", new Info(BreakInfo::pickaxe(0.4))));
+		self::register("warped_nylium", new GrassNylium(new BID(Ids::WARPED_NYLIUM), "Warped Nylium", new Info(BreakInfo::pickaxe(0.4))));
```
Comment (ShockedPlot7560) [inline src/block/VanillaBlocks.php]: Can be good to store in an variable the break info

File: src/block/GrassNylium.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\block\utils\BlockEventHelper;
+use pocketmine\item\Item;
+
+class GrassNylium extends Opaque{
```
Comment (ipad54) [inline src/block/GrassNylium.php]: `Nylium` or `NetherNylium` would be better.

**Comment by ipad54** [comment]
Bonemeal logic is missing.

Also, this PR should target `minor-next` branch, not `stable`.

**Comment by ShockedPlot7560** [comment]
There is no block available for bone meal spreading, roots and fungus is missing

**Comment by ipad54** [comment]
> roots and fungus is missing

Roots are implemented. https://github.com/pmmp/PocketMine-MP/blob/stable/src/block/NetherRoots.php

File: src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php [hunk 6]
```diff
@@ -983,6 +984,7 @@ private function registerSimpleDeserializers() : void{
 		$this->mapSimple(Ids::TUFF_BRICKS, fn() => Blocks::TUFF_BRICKS());
 		$this->mapSimple(Ids::UNDYED_SHULKER_BOX, fn() => Blocks::SHULKER_BOX());
 		$this->mapSimple(Ids::WARPED_WART_BLOCK, fn() => Blocks::WARPED_WART_BLOCK());
+		$this->mapSimple(Ids::WARPED_NYLIUM, fn() => Blocks::WARPED_NYLIUM());
```
Comment (dktapps) [inline src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php:986]: Alphabetical order please

File: src/data/bedrock/block/convert/BlockObjectToStateSerializer.php [hunk 7]
```diff
@@ -1060,6 +1061,7 @@ private function registerSimpleSerializers() : void{
 		$this->mapSimple(Blocks::TUFF(), Ids::TUFF);
 		$this->mapSimple(Blocks::TUFF_BRICKS(), Ids::TUFF_BRICKS);
 		$this->mapSimple(Blocks::WARPED_WART_BLOCK(), Ids::WARPED_WART_BLOCK);
+		$this->mapSimple(Blocks::WARPED_NYLIUM(), Ids::WARPED_NYLIUM);
```
Comment (dktapps) [inline src/data/bedrock/block/convert/BlockObjectToStateSerializer.php:1063]: Same again here

File: src/world/generator/object/NetherGrass.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+use pocketmine\block\BlockTypeIds;
+use pocketmine\block\VanillaBlocks;
+use pocketmine\math\Vector3;
+use pocketmine\utils\Random;
+use pocketmine\world\ChunkManager;
+use function count;
+
+class NetherGrass{
+	public static function growGrass(ChunkManager $world, Vector3 $pos, Random $random, int $count = 8, int $radius = 5) : void{
```
Comment (ShockedPlot7560) [inline src/world/generator/object/NetherGrass.php:34]: As there is no nether generator, I don't see the point of including a single function in a new class in the generator objects.

File: src/block/RedMushroom.php (hunk 3 above)
Comment (ShockedPlot7560) [reply src/block/RedMushroom.php:30]: This isn't resolved, you still add the use of StaticSupportTrait

File: src/block/NetherNylium.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+	public function ticksRandomly() : bool{
+		return true;
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		$world = $this->position->getWorld();
+		if($item instanceof Fertilizer){
+			$item->pop();
+			NetherGrass::growGrass($world, $this->position, new Random(mt_rand()), 8, 2);
```
Comment (ShockedPlot7560) [inline src/block/NetherNylium.php:52]: What's the point of integrating all the logic into a specific function that's only used once?

**Review by ShockedPlot7560 — CHANGES_REQUESTED** [review]

**Comment by dktapps** [comment]
Closing due to lack of activity

**Pull request closed without merging**